vkubelet/podcontroller.go

```diff
@@ -0,0 +1,380 @@
+}
+
+// syncPodInProvider tries and reconciles the state of a pod by comparing its Kubernetes representation and the provider's representation.
+func (pc *PodController) syncPodInProvider(ctx context.Context, pod *corev1.Pod) error {
+	// Create a span for the current call.
+	ctx, span := trace.StartSpan(ctx, "syncPodInProvider")
+	defer span.End()
+
+	// Add the pod's attributes to the current span.
+	addPodAttributes(span, pod)
+
+	// Check whether the pod has been marked for deletion.
+	// If it does, delete it in the provider.
+	if pod.DeletionTimestamp != nil {
+		// Delete the pod.
+		if err := pc.server.deletePod(ctx, pod.Namespace, pod.Name); err != nil {
+			err := pkgerrors.Wrapf(err, "failed to delete pod %q in the provider", loggablePodName(pod))
+			span.SetStatus(ocstatus.FromError(err))
+			return err
+		}
+		return nil
+	}
+
+	// Ignore the pod if it is in the "Failed" state.
+	if pod.Status.Phase == corev1.PodFailed {
+		log.G(ctx).Warnf("skipping sync of pod %q in %q phase", loggablePodName(pod), pod.Status.Phase)
+		return nil
+	}
+
+	// Create or update the pod in the provider.
+	if err := pc.server.createOrUpdatePod(ctx, pod); err != nil {
+		err := pkgerrors.Wrapf(err, "failed to sync pod %q in the provider", loggablePodName(pod))
+		span.SetStatus(ocstatus.FromError(err))
+		return err
+	}
+	return nil
+}
+
+// deleteDanglingPods checks whether the provider knows about any pods which Kubernetes doesn't know about, and deletes them.
+func (pc *PodController) deleteDanglingPods() {
+	// Create a span for the current call.
+	ctx, span := trace.StartSpan(pc.context, "deleteDanglingPods")
+	defer span.End()
+
+	// Grab the list of pods known to the provider.
+	pps, err := pc.server.provider.GetPods(ctx)
+	if err != nil {
+		err := pkgerrors.Wrap(err, "failed to fetch the list of pods from the provider")
+		span.SetStatus(ocstatus.FromError(err))
+		log.G(ctx).Error(err)
+		return
+	}
+
+	// Create a slice to hold the pods we will be deleting from the provider.
+	ptd := make([]*corev1.Pod, 0)
+
+	// Iterate over the pods known to the provider, marking for deletion those that don't exist in Kubernetes.
+	// Take on this opportunity to populate the list of key that correspond to pods known to the provider.
+	for _, pp := range pps {
+		if _, err := pc.podsLister.Pods(pp.Namespace).Get(pp.Name); err != nil {
+			if errors.IsNotFound(err) {
+				// The current pod does not exist in Kubernetes, so we mark it for deletion.
+				ptd = append(ptd, pp)
+				continue
+			}
+			// For some reason we couldn't fetch the pod from the lister, so we propagate the error.
+			err := pkgerrors.Wrap(err, "failed to fetch pod from the lister")
+			span.SetStatus(ocstatus.FromError(err))
+			log.G(ctx).Error(err)
+			return
+		}
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(ptd))
+
+	// Iterate over the slice of pods to be deleted and delete them in the provider.
+	for _, pod := range ptd {
+		go func(ctx context.Context, pod *corev1.Pod) {
```

Comment:
Can we use a semaphore to limit the concurrency here?
No need for anything fancy, can use a buffered channel to handle that.